File: Cpp/451_Sort_Characters_By_Frequency.cpp
class Solution {
public:
    string frequencySort(string s) {
        string result = "";
        unordered_map<char, int> frequency;
        for(auto& ch : s){
            ++frequency[ch];
        }
        
        priority_queue<pair<int, char>> pq;
        
        for(auto& it : frequency){
            pq.push({it.second, it.first});
        }
        
        while(!pq.empty()){
            auto current = pq.top();
            for(int i = current.first; i>0; --i){
                result += current.second;
            }
            pq.pop();
        }
        
        return result;
    }
};

File: Kotlin/451_Sort_Characters_By_Frequency.kt
class Solution {
    fun frequencySort(s: String): String {
        val result = StringBuilder()
        val m = hashMapOf<Char, Int>()
        for(i in s.indices) {
            m[s[i]] = m.getOrDefault(s[i], 0) + 1
        }

        val pq = PriorityQueue<Pair<Int, Char>>( compareByDescending{it.first} )
        m.forEach { (char, count) ->
            pq.add(Pair(count, char))
        }
        
        while(pq.isNotEmpty()) {
            val current = pq.remove()
            for(i in 0 until current.first){
                result.append(current.second)
            }
        }
        
        return result.toString()
    }
}